### DISCUSSION TITLE
Setting Cookies with Next Safe Action

### DISCUSSION THREAD

**Discussion by laurent512** [next-safe-action/next-safe-action#234] [General]
Hello, 

I tried to do set the cookies in an action and it does not work 

I have the following error :  Action error: Cookies can only be modified in a Server Action or Route Handler. Read more: https://nextjs.org/docs/app/api-reference/functions/cookies#cookiessetname-value-options

Example of code 
```
export const getAccess = propertyAction
    .metadata({
        action: "getAccess",
    })
    .action(async ({ ctx: { propertyMemberId, propertyId }, parsedInput }) => {
        cookies().set({
            name: "test",
            value: "test",
            httpOnly: true,
            path: "/",
        });
        return true;
    });
```

Is it expected? How would you be able to set a cookie diferently?

**Comment by TheEdoRan** [comment]
> Is it expected?

Not at all, on my end the cookie gets correctly set with your exact same code, both via direct execution and hooks. Can you please provide a link to a minimal reproduction of this issue? Thanks.

**Reply by laurent512** [reply]
I think it happens because I don't trigger it with a click of a button or a form. 
The way I use it is a bit of an anti pattern, I included it in an async server component to get dynamic data before the page is delivered.

In that case , it seems it is not perceived by nextjs as a server action (but this is not depending on nsa) so we can close I think

**Reply by TheEdoRan** [reply]
So you're trying to set the cookie while the HTTP is streaming in? If that's the case, unfortunately there's no way to do it at the page/component level, I tried to make it work in every way possible in the past, with no success, because that's a streaming limitation. The only place where you can do it right now is in `middleware.ts`, which is very limited by the edge runtime. The good news is that the Next.js team is apparently working on route-level middleware to solve the problem: 
- https://github.com/vercel/next.js/discussions/46722#discussioncomment-10262088

I know that's kinda off topic but I also know the pain very well, hope this helps.
